File: ctsa-capstone-microservices/career-trend-service/src/main/java/io/ctsa/careertrendservice/service/SalaryService.java
package io.ctsa.careertrendservice.service;

import io.ctsa.careertrendservice.model.Salary;
import io.ctsa.careertrendservice.prediction.ExponentialSmoothingFormula;
import io.ctsa.careertrendservice.repository.SalaryRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class SalaryService {

    private final SalaryRepository salaryRepository;

    private final ExponentialSmoothingFormula exponentialSmoothingFormula;

    public SalaryService(SalaryRepository salaryRepository,
                         ExponentialSmoothingFormula exponentialSmoothingFormula) {
        this.salaryRepository = salaryRepository;
        this.exponentialSmoothingFormula = exponentialSmoothingFormula;
    }

    public List<Salary> getAll() {
        return salaryRepository.findAll();
    }

    public Optional<Salary> getById(int salaryId) {
        return salaryRepository.findById(salaryId);
    }

    public List<Salary> getAllByYear(int year) {
        return salaryRepository.findAllByYear(year);
    }

    public Salary predict(int predictedYear, int majorId) {
        Salary nearestPredictionModel = salaryRepository.findFirstByMajorIdOrderByYearDesc(majorId);

        Salary predictionModel = new Salary();
        predictionModel.setMajorId(majorId);
        predictionModel.setYear(predictedYear);

        predictionModel = exponentialSmoothingFormula.predict(nearestPredictionModel, predictionModel);

        return predictionModel;
    }

    public Salary getLatest(int majorId) {
        return salaryRepository.findFirstByMajorIdOrderByYearDesc(majorId);
    }
}

File: ctsa-capstone-microservices/career-trend-service/src/main/java/io/ctsa/careertrendservice/service/SalaryService.kt
package io.ctsa.careertrendservice.service

import io.ctsa.careertrendservice.model.Salary
import io.ctsa.careertrendservice.prediction.storage.SmoothingParams
import io.ctsa.careertrendservice.prediction.storage.SmoothingParamsConstants.*
import io.ctsa.careertrendservice.prediction.timeseries.ExponentialSmoothingFormula
import io.ctsa.careertrendservice.repository.SalaryRepository
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.Optional

@Service
class SalaryService(private val salaryRepository: SalaryRepository,
                    private val exponentialSmoothingFormula: ExponentialSmoothingFormula,
                    private val smoothingParamsService: SmoothingParamsService) {

    companion object {
        private val log = LoggerFactory.getLogger(SalaryService::class.java)
    }

    val all: List<Salary>
        get() = salaryRepository.findAll()

    fun getById(salaryId: Int): Optional<Salary> {
        return salaryRepository.findById(salaryId)
    }

    fun getAllByYear(year: Int): List<Salary> {
        return salaryRepository.findByYear(year)
    }

    fun predict(predictedYear: Int, majorId: Int): Salary {
        val smoothingParams = smoothingParamsService.getSmoothingParams("$SALARY-$majorId")

        val nearestPredictionModel = salaryRepository.findFirstByMajorIdOrderByYearDesc(majorId)

        var predictionModel = Salary()
        predictionModel.majorId = majorId
        predictionModel.year = predictedYear

        predictionModel = exponentialSmoothingFormula.predict(nearestPredictionModel, predictionModel,
                                                              smoothingParams.alpha, smoothingParams.beta)

        return predictionModel
    }

    fun getLatest(majorId: Int): Salary {
        return salaryRepository.findFirstByMajorIdOrderByYearDesc(majorId)
    }

    fun estimateSmoothingParams(majorId: Int): SmoothingParams {
        log.info("Estimating smoothing parameters for Salary with major-id = $majorId")
        val salaries = salaryRepository.findByMajorIdOrderByYearAsc(majorId)

        return smoothingParamsService.findSmoothingParams("$SALARY-$majorId",
                                                          exponentialSmoothingFormula,
                                                          salaries)
    }
}

